src/lib/RoutingManager.ts

```diff
@@ -35,139 +32,45 @@ class RoutingManager implements Widget {
-      });
-
-      if (isEqual(nextUiState, widgetsUiState)) {
-        return;
-      }
-
-      this.currentUiState = nextUiState;
-
-      const searchParameters = this.getAllSearchParameters({
-        currentSearchParameters: state,
-        uiState: this.currentUiState,
-      });
-
-      helper
-        .overrideStateWithoutTriggeringChangeEvent(searchParameters)
-        .search();
-    });
+    walk(this.instantSearchInstance.mainIndex, current => {
+      const widgets = current.getWidgets();
+      const indexUiState = currentUiState[current.getIndexId()] || {};
 
-    this.renderURLFromState = event => {
-      this.currentUiState = this.getAllUiStates({
-        searchParameters: event.state,
-      });
-
-      const route = this.stateMapping.stateToRoute(this.currentUiState);
+      const searchParameters = widgets.reduce((parameters, widget) => {
+        if (!widget.getWidgetSearchParameters) {
+          return parameters;
+        }
 
-      this.router.write(route);
-    };
+        return widget.getWidgetSearchParameters(parameters, {
+          uiState: indexUiState,
+        });
+      }, current.getHelper()!.state);
 
-    helper.on('change', this.renderURLFromState);
+      current
+        .getHelper()!
+        .overrideStateWithoutTriggeringChangeEvent(searchParameters);
 
-    // Compare initial state and first render state to see if the query has been
-    // changed by the `searchFunction`. It's required because the helper of the
-    // `searchFunction` does not trigger change event (not the same instance).
-    const firstRenderState = this.getAllUiStates({
-      searchParameters: state,
+      this.instantSearchInstance.scheduleSearch();
     });
 
-    if (!isEqual(this.initState, firstRenderState)) {
-      // Force update the URL, if the state has changed since the initial read.
-      // We do this in order to make the URL update when there is `searchFunction`
-      // that prevents the search of the initial rendering.
-      // See: https://github.com/algolia/instantsearch.js/issues/2523#issuecomment-339356157
-      this.currentUiState = firstRenderState;
-
-      const route = this.stateMapping.stateToRoute(this.currentUiState);
-
-      this.router.write(route);
-    }
-  }
-
-  public getConfiguration(
-    currentConfiguration: SearchParameters
-  ): SearchParameters {
-    return this.getAllSearchParameters({
-      uiState: this.currentUiState,
-      currentSearchParameters: currentConfiguration,
-    });
+    // @TODO: Update state on external route update (popState)
+    // this.router.onUpdate(route => {});
   }
 
-  public init({ state }: { state: SearchParameters }): void {
-    // Store the initial state from the storage to compare it with the state on next renders
-    // in case the `searchFunction` has modified it.
-    this.initState = this.getAllUiStates({
-      searchParameters: state,
-    });
-  }
+  public write({ state }: { state: UiState }) {
```

Comment:
For signature consistency:

```suggestion
  public write({ state }: { state: UiState }): void {
```